freedomofpress/pressfreedomtracker.us
# Add fixture for geonames of cities with population > 500

## Description

Fixes #1731 

Changes proposed in this pull request:

* Adds new fixtures for geonames to include US cities with populations of at least 500.

## Type of change



## Testing

1. Load the fixtures with:

```
docker compose exec django python manage.py loaddata --app geonames cities500-us-only.json.xz
```

This should complete without errors and there should be 21149 rows in `geonames_geoname`.

2. Update an incident to have a city value of `Marion` and a state value of `Kansas`. This should update the coordinates on that incident to:
```
latitude  │ 38.34835
longitude │ -97.01725
```

whereas, without these fixtures, they would both be null.

3. Then, the incident should display in a bubble map chart. One way to artificially contrive this situation is to add a new tag to the incident you changed in step 2, then add a bubble map chart to a blog page and limit the dataset to that tag. The incident should appear in Kansas (and at a different location if "Group by State" is selected).

### Post-deployment actions

Load the fixtures in production with the management command:

```
python manage.py loaddata --app geonames cities500-us-only.json.xz
```
## Checklist

### General checks


## How I created these fixtures

1. Navigate to the tracker working directory and start the containers: `docker compose up`
2. Download cities500.zip from http://download.geonames.org/export/dump/ -- this produces cities500.txt in your tracker working directory.
3. Connect to the running tracker postgresql DB with `psql` (there are a number of ways to do this, let me know if you need help with this).
4. Create a temporary holding space for the "raw" cities500.txt data:
  ```psql
create schema if not exists raw;

create table raw.geonames
 (
   geonameid         bigint,
   name              text,
   asciiname         text,
   alternatenames    text,
   latitude          double precision,
   longitude         double precision,
   feature_class     text,
   feature_code      text,
   country_code      text,
   cc2               text,
   admin1_code       text,
   admin2_code       text,
   admin3_code       text,
   admin4_code       text,
   population        bigint,
   elevation         bigint,
   dem               bigint,
   timezone          text,
   modification      date
 );
```
5. A nice way to import data from a CSV file is the tool [`pgloader`](https://pgloader.readthedocs.io/en/latest/index.html). There are multiple ways to install this, but I used docker: `docker pull ghcr.io/dimitri/pgloader:latest`
6. Load the "raw" data into the database:
```
docker run --rm -it \
  --add-host host.docker.internal:host-gateway \
  -v .:/data \
  ghcr.io/dimitri/pgloader:latest pgloader \
  --type csv \
  /data/cities500.txt \
  postgresql://tracker:trackerpassword@host.docker.internal:5432/trackerdb?tablename=raw.geonames
```
7. Remove old data from the geonames table (this is not strictly necessary, but it doesn't hurt): `delete from geonames_geoname where 1 = 1;`
8. Copy the US cities from the "raw" data into our geonames database:
```sql
insert into geonames_geoname
  with geo as
  (
     select geonameid,
            name,
            latitude, 
            longitude,
            country_code,
            admin1_code,
            admin2_code,
            feature_class,
            feature_code,
            population,
            elevation,
            timezone
       from raw.geonames
       where country_code = 'US'
   )
     select geo.geonameid,
            geo.name,
            geo.latitude,
            geo.longitude,
            region.regcode,
            district.discode,
            population,
            elevation,
            country.isocode
       from geo
            left join geonames_country country
              on country.iso = geo.country_code

            left join geonames_region region
              on region.isocode = country.isocode
             and region.regcode = geo.admin1_code

            left join geonames_district district
              on district.isocode = country.isocode
             and district.regcode = geo.admin1_code
             and district.discode = geo.admin2_code;
```
If you did not remove old data, you may need to add an [`on conflict do nothing` clause](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT) to the end of the above query.
9. Dump the geonames data as a fixture:
```
docker compose exec django ./manage.py dumpdata geonames -o cities500-us-only.json.xz
```
